security/pkg/pki/ca/selfsignedcarootcertrotator.go

```diff
@@ -0,0 +1,210 @@
+			rootCertRotatorLog.Warn("CA cert in KeyCertBundle does not match CA cert in " +
+				"istio-ca-secret. Start to reload root cert into KeyCertBundle")
+			if err := rotator.ca.GetCAKeyCertBundle().VerifyAndSetAll(caSecret.Data[caCertID],
+				caSecret.Data[caPrivateKeyID], nil, caSecret.Data[caCertID]); err != nil {
+				rootCertRotatorLog.Errorf("failed to reload root cert into KeyCertBundle (%v)", err)
+			}
+			rootCertRotatorLog.Info("Successfully reloaded root cert into KeyCertBundle.")
+			certEncoded := base64.StdEncoding.EncodeToString(rotator.ca.GetCAKeyCertBundle().GetRootCertPem())
+			if err = rotator.configMapController.InsertCATLSRootCertWithRetry(
+				certEncoded, rotator.config.retryInterval, 30*time.Second); err != nil {
+				rootCertRotatorLog.Errorf("Failed to write self-signed Citadel's root cert "+
+					"to configmap (%s). Node agents will not be able to connect.",
+					err.Error())
+			}
+			rootCertRotatorLog.Info("Root cert is updated into configmap.")
+		}
+		return
+	}
+
+	rootCertRotatorLog.Infof("Refresh root certificate, root cert is about to expire: %s", err.Error())
+	options := util.CertOptions{
+		TTL:           rotator.config.caCertTTL,
+		SignerPrivPem: caSecret.Data[caPrivateKeyID],
+		Org:           rotator.config.org,
+		IsCA:          true,
+		IsSelfSigned:  true,
+		RSAKeySize:    caKeySize,
+		IsDualUse:     rotator.config.dualUse,
+	}
+	pemCert, pemKey, ckErr := util.GenRootCertFromExistingKey(options)
+	if ckErr != nil {
+		rootCertRotatorLog.Errorf("unable to generate CA cert and key for self-signed CA: %s", ckErr.Error())
+		return
+	}
+
+	oldCaCert := caSecret.Data[caCertID]
+	oldCaPrivateKey := caSecret.Data[caPrivateKeyID]
+	caSecret.Data[caCertID] = pemCert
+	caSecret.Data[caPrivateKeyID] = pemKey
+	if err = rotator.caSecretController.UpdateCASecretWithRetry(caSecret, rotator.config.retryInterval, 30*time.Second); err != nil {
+		rootCertRotatorLog.Errorf("Failed to write secret to CA secret (error: %s). "+
+			"Abort new root certificate.", err.Error())
+		return
+	}
+	rootCertRotatorLog.Infof("A new self-generated root certificate is written into secret: %v", string(pemCert))
+
+	if err := rotator.ca.GetCAKeyCertBundle().VerifyAndSetAll(pemCert, pemKey, nil, pemCert); err != nil {
+		rootCertRotatorLog.Errorf("Failed to create CA KeyCertBundle (%s)", err.Error())
+		return
+	}
+	rootCertRotatorLog.Infof("Updated CA KeyCertBundle using existing public key: %v", string(pemCert))
+
+	certEncoded := base64.StdEncoding.EncodeToString(rotator.ca.GetCAKeyCertBundle().GetRootCertPem())
+	if err = rotator.configMapController.InsertCATLSRootCertWithRetry(
+		certEncoded, rotator.config.retryInterval, 30*time.Second); err != nil {
+		rootCertRotatorLog.Errorf("Failed to write self-signed Citadel's root cert "+
+			"to configmap (%s). Rollback root cert.", err.Error())
+		caSecret.Data[caCertID] = oldCaCert
+		caSecret.Data[caPrivateKeyID] = oldCaPrivateKey
```

Comment:
I think you may need to read the secret and then update. I'm not sure this will succeed, because the ca secret is not the same as you read the caSecret.

Reply:
Good catch. I have verified that the old secret cannot be used to roll back root certificate in istio-ca-secret. I have added one step that loads istio-ca-secret for rollback.